hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastOperandTypes.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.calcite.validate.types;
+
+import org.apache.calcite.rel.type.RelDataTypeComparability;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlOperandCountRange;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.type.ComparableOperandTypeChecker;
+import org.apache.calcite.sql.type.OperandTypes;
+import org.apache.calcite.sql.type.SqlOperandTypeChecker;
+import org.apache.calcite.sql.type.SqlSingleOperandTypeChecker;
+
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.typeName;
+import static org.apache.calcite.sql.type.SqlTypeName.ANY;
+import static org.apache.calcite.sql.type.SqlTypeName.NULL;
+
+/**
+ * A collection of operand type checkers. Basically, a mirror of {@link
+ * OperandTypes} provided by Calcite with various enhancements.
+ */
+public final class HazelcastOperandTypes {
+
+    /**
+     * The same as Calcite's {@link OperandTypes#COMPARABLE_ORDERED_COMPARABLE_ORDERED},
+     * but selects the least restrictive type as a common type. We do character
+     * coercion provided by {@link SqlOperandTypeChecker.Consistency#COMPARE} and
+     * used by Calcite on our own.
+     */
+    public static final SqlOperandTypeChecker COMPARABLE_ORDERED_COMPARABLE_ORDERED =
+            new ComparableOperandTypeChecker(2, RelDataTypeComparability.ALL,
+                    SqlOperandTypeChecker.Consistency.LEAST_RESTRICTIVE);
+
+    private HazelcastOperandTypes() {
+    }
+
+    /**
+     * @return the base operand type checker wrapped into a new type checker
+     * disallowing ANY type.
+     */
+    public static SqlSingleOperandTypeChecker notAny(SqlOperandTypeChecker base) {
+        return new NotAny(base);
+    }
+
+    /**
+     * @return the base operand type checker wrapped into a new type checker
+     * disallowing all of the operands to be of NULL type simultaneously.
+     */
+    public static SqlOperandTypeChecker notAllNull(SqlOperandTypeChecker base) {
```

Comment:
We do we disallow all-null operands? I see that the same behavior is observed in, e.g., PostgreSQL. Is it possible to explain the reason of this decision in the design doc?

Reply:
There is a conceptual problem with NULLs and operator overloads: NULL inherently doesn't have any type, so it's impossible to select an overload unambiguously if all operands are NULL. And there is a technical problem: Calcite expects that every NULL literal participating in expression should have a type assigned (NULL type doesn't work here), so they use the same approach as PG. I have tried to workaround that and make NULL type a first class citizen, but that produced enormous amount of strange artifacts in Calcite type inference. As far as I remember, it was something like `NULL + NULL -> NULL`, but at the same time `NULL + NULL + NULL -> DECIMAL`, `ABS(NULL + NULL)` was producing an error while `ABS(NULL)` was fine, etc. So I decided to give up, especially considering that queries like `NULL + NULL` are pretty rare in practice.